### DISCUSSION TITLE
Dynamically update Tailwind theme values based on color mode and density

### DISCUSSION THREAD

**Discussion by thedanchez** [tailwindlabs/tailwindcss#10633] [Help]
I think there may have been things written on this already but I'm having a hard time finding the material to get a clear answer to this question. I would like to use TailwindCSS to build a component library (in SolidJS -- not that it necessarily matters) and there's this use-case that I have in mind that I don't know if it supports. 

The breakdown is the following. There are two axes that my theme leans on:

```typescript
// your typical light vs dark mode
export type ColorMode = "light" | "dark";

// the density value affects spacing, font-sizes, line-height, etc. across all custom components
export type Density = "narrow" | "regular" | "wide"; 
```

The key thing to focus is on the `Density` dimension of the theme. I know TailwindCSS supports updating color values around `light` vs `dark` mode, but I'm curious if it supports dynamically updating the theme `spacing` values if the user decides to change their density mode? I'm sure I can get something like this working if I was using something like say Theme UI or some CSS-in-JS solution to dynamically update my theme object and propagate the changes downstream but I would like to lean into more native-like CSS solutions like Tailwind. Another option that I was exploring that I think could work for this is something like [Sprinkles](https://vanilla-extract.style/documentation/packages/sprinkles/).

Any help is appreciated!

Side Note: If there is a way to do this, I think it would be extremely useful to have this documented on the main TailwindCSS site so that others can see an example of this aside from the typical light vs dark theme examples

**Comment by crswll** [comment]
There's a few ways to approach this.

## Multiplier

You can use a multiplier in the `spacing` values for your config (and other values where it makes sense) but when you need to fine-tune something for a specific density and/or breakpoint it gets to be kind of a nightmare. A quick code sample of what I mean:

```js
module.exports = {
  theme: {
    spacing: {
      small: 'calc(var(--spacing-base) * 1)',
      medium: 'calc(var(--spacing-base) * 2)',
      large: 'calc(var(--spacing-base) * 3)',
    },
  },
}
```

## Theme Composing

I wrote a plugin which gives you something more like the Sprinkles approach. You essentially provide multiple `themes` and they're activated based on selectors or media queries. So you can do something like:

```js
const baseTheme = {
  // Any theme based stuff can go in here, just make sure `base`
  // has default values for any values you may add...
  spacing: {
    small: '3rem',
    medium: '5rem',
    large: '7rem',
  }
}

const wideTheme = {
  spacing: {
    small: '4rem',
    medium: '6rem',
    large: '8rem',
  },
}

const narrowTheme = {
  spacing: {
    small: '2rem',
    medium: '4rem',
    large: '6rem',
  },
}

module.exports = {
  plugins: [
    themeSwapper({
      themes: [
        {
          name: 'base',
          selectors: [':root'],
          theme: baseTheme,
        },
        {
          name: 'wide',
          selectors: ['.wide'],
          theme: wideTheme,
        },
        {
          name: 'narrow',
          selectors: ['.narrow'],
          theme: narrowTheme,
        },
      ],
    }),
  ],
}
```

Hacked up a little demo for this here: https://play.tailwindcss.com/7gJjDLYslk?file=config

## Just Logic with a "variants" like API.

You can use something like https://github.com/joe-bell/cva or https://github.com/crswll/clb to code up something like:

```js
const clb = require('clb')

const buttonBuilder = clb({

  /* This can be anything `classnames` accepts. */
  base: 'font-serif rounded-2xl',

  defaultVariants: {
    color: 'gray',
    size: 'medium',
    spacing: 'medium',
  },

  /*
    The value for each variant value below can be anything
    `classnames` accepts.
  */
  variants: {
    color: {
      gray: 'text-gray-800 bg-gray-800',
      red: 'text-red-800 bg-red-200',
      blue: 'text-blue-800 bg-blue-200',
      green: 'text-green-800 bg-green-200',
    },
    size: {
      small: 'text-sm',
      medium: 'text-md',
      large: 'text-lg',
    },
    spacing: {
      small: 'p-2',
      medium: 'p-4',
      large: 'p-6',
    },
  },
})

buttonBuilder()
// -> font-serif rounded-2xl text-gray-800 bg-gray-800 text-md p-4

buttonBuilder({ color: 'red' })
// -> font-serif rounded-2xl text-red-800 bg-red-800 text-md p-4

buttonBuilder({ color: 'blue', size: 'large' })
// -> font-serif rounded-2xl text-blue-800 bg-blue-800 text-lg p-6
```

Will clean this up in the morning but think there's plenty to explore here for now.
